### PR TITLE
Remove python 3.7, add python3.11

### PR DESCRIPTION
**Pull request by albi3ro**
This PR drops Python 3.7 support and adds Python 3.11 support.

Python 3.7 is no longer supported by major packages, see [NEP 29 — Recommend Python and NumPy version support as a community policy standard](https://numpy.org/neps/nep-0029-deprecation_policy.html).

To accomplish adding 3.11 support, we need to:

* Bump the CI version of torch to `1.13.0` from `1.11.0+cpu`. This required a test fix due to a type comparison issue.
* Drop `tensornetwork` from the requirements (no longer used by anything)
* Only install `cvxopt` and `cvxpy` if the python version isn't Python 3.11.   `cvxopt` doesn't have Python 3.11 wheels
* Make `opt_einsum` a qcut-tests-only requirement.  Doesn't have gpu compatible wheels for Python 3.11
* Avoid mocking a mock object. Python 3.11 must have changed something so we can't do that anymore.

We would also not test tensorflow on 3.11, since no wheels exist for tensorflow with Python 3.11.

Hanging 3.7 tests are because they are required tests as specified in our repository settings.  After this PR is merged, I will update our repository settings.

### CODE REVIEW

File: tests/ops/qubit/test_parametric_ops.py [hunk 1]
```diff
@@ -2905,7 +2905,7 @@ def test_pauli_rot_identity_torch(self, torch_device, theta):
 
         val = np.cos(-theta / 2) + 1j * np.sin(-theta / 2)
         exp = torch.tensor(np.diag([val, val]), device=torch_device)
-        assert torch.allclose(mat, exp)
+        assert qml.math.allclose(mat, exp)
```
Comment (albi3ro) [inline tests/ops/qubit/test_parametric_ops.py:2908]: Was getting "`ComplexDouble` is not equal to `ComplexFloat`" errors with v1.13. `qml.math` ignores the typing.

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -8,7 +8,7 @@ on:
 
 env:
   TF_VERSION: 2.10.0
-  TORCH_VERSION: 1.11.0+cpu
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply .github/workflows/tests.yml:11]: I can't remember exactly the problem, but I remember it being a massive headache.  One of those "it doesn't work and I don't know why" turning into "it works and I don't know why" sort of things.  But doesn't seem to be an issue now.

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -19,22 +19,20 @@ jobs:
...
         config:
-          - {python-version: 3.7, suite: 'torch'}
-          - {python-version: 3.7, suite: 'tf'}
           - {python-version: 3.8, suite: 'tf'}
           - {python-version: '3.10', suite: 'tf'}
           - {python-version: 3.8, suite: 'torch'}
-          - {python-version: '3.10', suite: 'torch'}
+          - {python-version: '3.11', suite: 'torch'}
           - {python-version: 3.8, suite: 'jax'}
-          - {python-version: '3.10', suite: 'jax'}
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply .github/workflows/tests.yml:29]: I was thinking of just testing the lowest-supported python version and the highest-supported python version.  If both work, the things in between *should* work too.

[the remark this answers is not in the document]
Comment (josh146) [reply .github/workflows/tests.yml:29]: >  I see we don't have every combination of version and interface so I'm guessing it's intentional

historical, but this was intentional a couple of years back, as we wanted to find a good middle ground between testing every combination and reducing time/energy of the CI

**Review by timmysilv — APPROVED** [review]
cool stuff

**Pull request merged**